// symmetry/box.js
function Box(centers) {
    if (centers) this.centers = centers;
    else this.centers = [0, 1, 2, 3, 4, 5];
}

Box.prototype.rotateY = function(yPower) {
    if (yPower == 0) return this;
    var newCenters = [this.centers[4], this.centers[5], this.centers[2], this.centers[3],
                      this.centers[1], this.centers[0]];
    return (new Box(newCenters)).rotateY(yPower - 1);
}

Box.prototype.flipLR = function(flag) {
    if (!flag) return this;
    var centers = this.centers.slice();
    centers[5] = this.centers[4];
    centers[4] = this.centers[5];
    return new Box(centers);
}

Box.prototype.rotateZ2 = function(flag) {
    if (!flag) return this;
    var centers = this.centers.slice();
    centers[3] = this.centers[2];
    centers[2] = this.centers[3];
    centers[5] = this.centers[4];
    centers[4] = this.centers[5];
    return new Box(centers);
}

Box.prototype.apply = function(aBox) {
    // applies aBox to this
    var newCenters = new Array(6);
    for (var i = 0; i < 6; i++) {
        newCenters[i] = this.centers[aBox.centers[i]];
    }
    return new Box(newCenters);
}

Box.prototype.equals = function(aBox) {
    for (var i = 0; i < 6; i++) {
        if (aBox.centers[i] != this.centers[i]) return false;
    }
    return true;
}

exports.Box = Box;
